File: Week_09/java/src/_438_FindAllAnagramsInAString.java
import java.util.*;

/**
 * 第一遍：2020/07/16周四 ✅
 * 第二遍：2020/07/15周三
 * 第三遍：2020/07/08周四
 * 第四遍：2020/07/13周一
 * 第三遍：2020/06/29周一
 * 第四遍：2020/07/05周日
 * 我的思路就是判断滑动窗口中的每一个字符出现的次数是否相同。题解的解法没看懂，花了两个小时。卧槽。
 */
class _438_FindAllAnagramsInAString {
    public List<Integer> findAnagrams(String s, String p) {
        if (s == null || p == null || p.length() > s.length()) return Collections.emptyList();
        int[] table = new int[26];
        for (int i = 0; i < p.length(); i++) {
            table[p.charAt(i) - 'a']++;
            table[s.charAt(i) - 'a']--;
        }
        List<Integer> ret = new ArrayList<>(s.length() - p.length() + 1);
        if (isAnagram(table)) ret.add(0);
        for (int i = p.length(); i < s.length(); i++) {
            table[s.charAt(i - p.length()) - 'a']++;
            table[s.charAt(i) - 'a']--;
            if (isAnagram(table)) ret.add(i - p.length() + 1);
        }
        return ret;
    }

    boolean isAnagram(int[] table) {
        for (int v : table) {
            if (v != 0) return false;
        }
        return true;
    }

    static class S1 {
        public List<Integer> findAnagrams11(String s, String p) {
            if (s == null || p == null || p.length() > s.length()) return Collections.emptyList();
            Map<Character, Integer> map = new HashMap<>();
            int length = p.length();
            for (int i = 0; i < length; i++) {
                map.put(p.charAt(i), map.getOrDefault(p.charAt(i), 0) + 1);
            }
            int counter = map.size();
            int left = 0, right = 0;
            List<Integer> ret = new ArrayList<>();
            while (right < s.length()) {
                char c = s.charAt(right);
                if (map.containsKey(c)) {
                    map.put(c, map.get(c) - 1);
                    if (map.get(c) == 0) counter--;
                }
                right++;
                while (counter == 0) {
                    char l = s.charAt(left);
                    if (map.containsKey(l)) {
                        if (map.get(l) == 0) counter++;
                        map.put(l, map.get(l) + 1);
                    }
                    if (right - left == length) ret.add(left);
                    left++;
                }
            }
            return ret;
        }

        //看不懂的解法一。。。昨天状态不好，今天很容易就看懂了
        public List<Integer> findAnagrams(String s, String t) {
            List<Integer> ret = new ArrayList<>();
            if (t.length() > s.length()) return ret;
            Map<Character, Integer> map = new HashMap<>();
            for (int i = 0; i < t.length(); i++) {
                char c = t.charAt(i);
                map.put(c, map.getOrDefault(c, 0) + 1);
            }
            int counter = map.size();
            int begin = 0, end = 0;
            while (end < s.length()) {
                char c = s.charAt(end);
                if (map.containsKey(c)) {
                    int count = map.get(c) - 1;
                    map.put(c, count);
                    if (count == 0) counter--;
                }
                end++;
                while (counter == 0) {
                    char temp = s.charAt(begin);
                    if (map.containsKey(temp)) {
                        int count = map.get(temp) + 1;
                        map.put(temp, count);
                        if (count > 0) counter++;
                    }
                    if (end - begin == t.length()) ret.add(begin);
                    begin++;
                }
            }
            return ret;
        }
    }

    public static void main(String[] args) {
        System.out.println(new S1().findAnagrams11("cbaebabacd", "abc"));
        System.out.println(new S1().findAnagrams("baa", "aba"));
//        System.out.println(new S2().findAnagrams("baa", "aba"));
        System.out.println(new S1().findAnagrams("bba", "aba"));
//        System.out.println(new S2().findAnagrams("bba", "aba"));
        System.out.println(new S2().findAnagrams("cbaebabacd", "abc"));
    }

    static class S2 {

        //看不懂的解法2
        public List<Integer> findAnagrams(String s, String p) {
            List<Integer> ret = new ArrayList<>();
            if (p.length() > s.length()) return ret;
            int[] table = new int[256];
            for (int i = 0; i < p.length(); i++) {
                table[p.charAt(i)]++;
            }
            int left = 0, right = 0, count = p.length();
            while (right < s.length()) {
                if (table[s.charAt(right++)]-- > 0) count--;
                if (count == 0) ret.add(left);
                if (right - left == p.length() && table[s.charAt(left++)]++ >= 0) count++;
            }
            return ret;
        }
    }
}

File: Week_09/kotlin/src/_438_FindAllAnagramsInAString.kt
import java.util.*
import kotlin.collections.ArrayList


class _FindAllAnagramsInAString_ {
    class Solution {
        fun findAnagrams(s: String, p: String): List<Int> {
            // 该滑动窗口算法，并没有将窗口的长度限制为p的长度
            // 而是先移动右端点，直到窗口中包含了p中所有的字符，记该条件为a，
            // 在条件a不变的情况下，移动左端点，减小窗口的长度，最终如果窗口的长度和p的长度相同
            // 则当前窗口中的字符满足题目要求，也就是p的字母异位词.
            if (p.length > s.length) return Collections.emptyList()
            // 统计p中每个字符出现的次数
            val freq = HashMap<Char, Int>()
            for (c in p) {
                freq[c] = freq.getOrDefault(c, 0) + 1
            }
            // 获得p中字符的个数
            var count = freq.size
            // 滑动窗口的左端点和右端点
            var left = 0
            var right = 0
            // 保存子串的起始索引
            val ret = ArrayList<Int>()
            while (right < s.length) {
                // 先移动右端点
                var c = s[right]
                // 如果字符出现在p中
                if (freq.containsKey(c)) {
                    // 将字符出现的次数-1
                    freq[c] = freq[c]!! - 1
                    // 如果字符出现的次数为0，说明[left, right]中c出现的次数和c在p中出现的次数相同
                    if (freq[c]!! == 0) count--
                }
                right++
                // 如果p中每个字符都在[left, right]中出现了
                while (count == 0) {
                    // 如果[left, right]中字符的个数和p中字符的个数相同，那么s[left, right]就是p的字母异位词
                    if (right - left == p.length) ret.add(left)
                    c = s[left]
                    // 将left指向的字符从窗口中移除
                    if (freq.containsKey(c)) {
                        // 如果字符c出现在p中，
                        // 如果当前c在[left, right]中出现的次数和c在p中出现的次数相同，
                        // 则p中的所有字符不会出现在[left, right]中，因此移动right
                        // 那么将c出现的次数增加
                        if (freq[c]!! == 0) count++
                        freq[c] = freq[c]!! + 1
                    }
                    left++
                }
            }
            return ret
        }
    }

    class SlideWindow {
        fun findAnagrams(s: String, p: String): List<Int> {
            if (p.length > s.length) return Collections.emptyList()
            val table = IntArray(26)
            for (c in p) {
                table[c - 'a']++
            }
            var count = p.length
            val ret = ArrayList<Int>()
            var left = 0;
            var right = 0;
            while (right < s.length) {
                if (table[s[right++] - 'a']-- > 0) count--
                if (count == 0) ret.add(left)
                if (right - left >= p.length && table[s[left++] - 'a']++ >= 0) count++
            }
            return ret
        }
    }

}